### DISCUSSION TITLE
Deprecating Python 2.x compatibility

### DISCUSSION THREAD

**Discussion by sylikc** [sylikc/pyexiftool#9] [Ideas]
I have a question ... do you think it's ok to move entirely to a Python 3.x and start using new libraries if they become available?  would that hurt users?

What's a good Python 3.x baseline to start at?  3.2?  3.4?

**Comment by csparker247** [comment]
Since Python 2 is [officially no longer supported](https://www.python.org/doc/sunset-python-2/), I think it's reasonable to target Python 3 going forward. You could tag the last version of pyexiftool supporting Python 2, and if you really want to be nice, maintain a bug fix branch for that version which could be installed from source. 

As far as which version to target, I feel like I generally don't see many packages targeting anything below 3.6. But that's a subjective assessment. macOS users generally use the Homebrew packages, which provide Python 3.7+. Debian Stretch (old stable) provides 3.5. Debian Buster (stable) provides 3.7. Ubuntu 18.04LTS provides 3.6. Ubuntu 20.04LTS provides 3.8. I don't know what Windows users do.

Given these deployments, I would say 3.6 or 3.7 seem like good options. I guess it also largely depends on what built-in's you intend to use.

EDIT: For myself, I target 3.6 for library code but use it across many environments.

**Reply by sylikc** [reply]
Wow, thanks for all this information.  Yeah, I tracked when Python 2 went EOL.  

The insight about what comes with what distribution is very helpful.  I personally use Windows myself as the primary desktop and while there is a Microsoft Store app, but I generally deploy the installation from Python.org .

I do run other Linux distros and I also see that many packages target 3.6.  Appreciate all that research into the different releases!

**Comment by ickc** [comment]
Python 2 users using the deprecated Python 2 can still use older version of pyexiftool which is deprecated.

So no, you don't need to worry about that.

Also note that pypy will support python2 indefinitely as pypy is built upon the Python2 language. So python2 is here to stay in some way. But people want to stay have ways to do it.

Often, extreme cases are supported only when you have the extra energy (it has a strategic tax.) It would be more important to make your own life easier in maintaining it to guarantee years later when you don't have the energy you can still maintain it.

**Reply by jangop** [reply]
> Python 2 users using the deprecated Python 2 can still use older version of pyexiftool which is deprecated.

This.

Considering that it is likely going to take a while before this project has been properly restructured, I would personally go for 3.9 because of the [extended type hinting](https://www.python.org/dev/peps/pep-0585).
